### DISCUSSION TITLE
Routes with balanced durations

### DISCUSSION THREAD

**Discussion by TeodoraBogdan** [google/or-tools#4148] [Routing (and legacy CP) questions]
Hello!

I am trying to get routes with balanced durations. 
For example: I have an input with 100 stops, 3 vehicles which leave from the same stop and return to it. Their maximum duration can be of 4 hours
In the solution I receive, the first route has 4 hours, the next one 2 hours and the last one 2 and a half.
What I want is that their duration to be around the same values (max duration*threshold)

What I tried is:
```cs
for (int i = 0; i < m_vehiclesSize; i++) {
  int64 endTime = endTimes[i] - (endTimes[i] - startTimes[i]) * 0.15;	//end route with maximum 15% before the maximum duration
  time_dimension.SetCumulVarSoftLowerBound(manager->GetEndIndex(i), endTime, 100);
}
```

This doesn't work if the vehicle starts the route after the set start time, so it will not have to wait before visiting the first stop (if the vehicle has the start time at 8:00, first stop has the tw:9:00-10:00, and the travel time between them is 30 min => the vehicle has to wait 30 minutes before visiting the stop.
In my case, I let the vehicle to start the route at 8:30 so it won't have any wait time `time_dimension.CumulVar(manager->GetStartIndex(i))->SetMin(startTimes[i]);` ).
So if I have 3 vehicles, their routes duration of maximum 4 hours and the start time is 8:00, but one starts at 8:00, another one at 8:30 and another one at 8:15, how can I force them to have routes of similar durations (not similar end times, as it is in the code above)?

And also, the code above forces the routes to have a duration between 3:24-4:00 hours, but if all the routes can have around 2 hours for example, they should be allowed to have it. As I said, their duration should be balanced, not the end time.

Thank you in advance!

**Comment by sschnug** [comment]
Consider [SetGlobalSpanCostCoefficient](https://github.com/google/or-tools/blob/82750ac12f1ee5354e1c7869894d9af3508778f2/ortools/constraint_solver/routing.h#L2730) which is probably one of the few general-purpose linear mechanisms here.

Apart from that, there is probably some quadratic upper-bound time-window formulation (balancing often needs nonlinearity). But non-linear objectives are kind of outliers in the routing-solver (imho) and this makes param-tuning sometimes more troublesome.

**Comment by Mizux** [comment]
create a second dimension, duration dimension instead of time dimension whose:
* start is forced at zero,
* use the same transit id than the one used with the time dimension
* add constraint so for any node slackVar of the time dimension == slackVar of the duration dimension

**Reply by TeodoraBogdan** [reply]
Something like this?
```cs	
	m_routing->AddDimensionWithVehicleTransitAndCapacity(timeEvaluatorIndices, m_stops->GetMaxWaitTime(), m_vehicles->GetEndTimes(), false, "Time");


	m_routing->AddDimensionWithVehicleTransitAndCapacity(timeEvaluatorIndices, m_stops->GetMaxWaitTime(), m_vehicles->GetEndTimes(), true, "Duration");
	operations_research::RoutingDimension* durationDimension = m_routing->GetMutableDimension("Duration");
	operations_research::RoutingDimension* timeDimension = m_routing->GetMutableDimension("Time");
	operations_research::Solver& solver = *m_routing->solver();

	for (int i = 0; i < manager->num_nodes(); i++)
	{
		int nodeIndex = manager->IndexToNode(i).value();
		solver.AddConstraint(solver.MakeEquality(durationDimension->SlackVar(nodeIndex), timeDimension->CumulVar(nodeIndex)));
	}

```

**Reply by Mizux** [reply]
```cs
solver.AddConstraint(solver.MakeEquality(durationDimension->SlackVar(nodeIndex), timeDimension->SlackVar(nodeIndex)));
```

you must equalize the slack var in both dimension not the slack with the cumul

**Reply by TeodoraBogdan** [reply]
yes, my bad!

```cpp
	m_routing->AddDimensionWithVehicleTransitAndCapacity(timeEvaluatorIndices, m_stops->GetMaxWaitTime(), m_vehicles->GetEndTimes(), false, "Time");


	m_routing->AddDimensionWithVehicleTransitAndCapacity(timeEvaluatorIndices, m_stops->GetMaxWaitTime(), m_vehicles->GetEndTimes(), true, "Duration");
	operations_research::RoutingDimension* durationDimension = m_routing->GetMutableDimension("Duration");
	operations_research::RoutingDimension* timeDimension = m_routing->GetMutableDimension("Time");
	operations_research::Solver& solver = *m_routing->solver();

	for (int i = 0; i < manager->num_nodes(); i++)
	{
		int nodeIndex = manager->IndexToNode(i).value();
		solver.AddConstraint(solver.MakeEquality(durationDimension->SlackVar(nodeIndex), timeDimension->SlackVar(nodeIndex)));
	}
```
So this should be ok, right?

**Reply by TeodoraBogdan** [reply]
I don't think it works as expected, because this is how the routes were:
![image](https://github.com/google/or-tools/assets/52992360/12514b95-cc9c-4d32-9fde-063dbddfc691)
And this is how they are now:
![image](https://github.com/google/or-tools/assets/52992360/bf091aa3-6eb7-432e-afe1-0d595126b28b)
The routes seem to be more unbalanced that they were

**Reply by TeodoraBogdan** [reply]
If for the time dimenstion the constraint is:
```cpp
m_routing->AddDimensionWithVehicleTransitAndCapacity(
  timeEvaluatorIndices,
  m_stops->GetMaxWaitTime(),
  m_vehicles->GetEndTimes(),
  false,
  "Time");
```
for the duration dimension the slack_max  should be 0 or m_stops->GetMaxWaitTime()?
```cpp
m_routing->AddDimensionWithVehicleTransitAndCapacity(
  timeEvaluatorIndices,
  0,
  m_vehicles->GetEndTimes(),
  true,
  "Duration");
```
or
```cpp
m_routing->AddDimensionWithVehicleTransitAndCapacity(
  timeEvaluatorIndices,
  m_stops->GetMaxWaitTime(),
  m_vehicles->GetEndTimes(),
  true,
  "Duration");
```

**Reply by Mizux** [reply]
if in Duration dimension, you force stack to be 0, then it would mean the duration dim only represent the "transit time" 
i.e. a route with 2h transit + 4h of waiting time is identical to a route of 2h transit + 0 waiting time.

**Reply by TeodoraBogdan** [reply]
So if I want to take the wait time into consideration, because I am interested in the entire duration (wait time included), i should put   m_stops->GetMaxWaitTime() as slack max, right?

**Reply by TeodoraBogdan** [reply]
They still seem to not be that balanced:
![image](https://github.com/google/or-tools/assets/52992360/dc7d02ba-41fb-4b1a-bafb-c53f33da9659)
